src/udev/udev-event.c

```diff
@@ -121,71 +121,75 @@ static const struct subst_map_entry map[] = {
+        const char *val = NULL;
         char *s = dest;
+        dev_t devnum;
+        int r;
 
         assert(entry);
 
         switch (entry->type) {
         case SUBST_DEVPATH:
-                l = strpcpy(&s, l, udev_device_get_devpath(dev));
+                r = sd_device_get_devpath(dev, &val);
+                if (r < 0)
+                        return r;
+                l = strpcpy(&s, l, val);
                 break;
         case SUBST_KERNEL:
-                l = strpcpy(&s, l, udev_device_get_sysname(dev));
+                r = sd_device_get_sysname(dev, &val);
+                if (r < 0)
+                        return r;
+                l = strpcpy(&s, l, val);
                 break;
         case SUBST_KERNEL_NUMBER:
-                if (udev_device_get_sysnum(dev) == NULL)
-                        break;
-                l = strpcpy(&s, l, udev_device_get_sysnum(dev));
+                r = sd_device_get_sysnum(dev, &val);
+                if (r < 0)
+                        return r == -ENOENT ? 0 : r;
+                l = strpcpy(&s, l, val);
                 break;
         case SUBST_ID:
-                if (event->dev_parent == NULL)
-                        break;
-                l = strpcpy(&s, l, udev_device_get_sysname(event->dev_parent));
-                break;
-        case SUBST_DRIVER: {
-                const char *driver;
-
-                if (event->dev_parent == NULL)
-                        break;
-
-                driver = udev_device_get_driver(event->dev_parent);
-                if (driver == NULL)
-                        break;
-                l = strpcpy(&s, l, driver);
+                if (!event->dev_parent)
+                        return 0;
+                r = sd_device_get_sysname(event->dev_parent->device, &val);
+                if (r < 0)
+                        return r;
+                l = strpcpy(&s, l, val);
                 break;
-        }
-        case SUBST_MAJOR: {
-                char num[UTIL_PATH_SIZE];
-
-                sprintf(num, "%u", major(udev_device_get_devnum(dev)));
-                l = strpcpy(&s, l, num);
+        case SUBST_DRIVER:
+                if (!event->dev_parent)
+                        return 0;
+                r = sd_device_get_driver(event->dev_parent->device, &val);
+                if (r < 0)
+                        return r == -ENOENT ? 0 : r;
+                l = strpcpy(&s, l, val);
                 break;
-        }
-        case SUBST_MINOR: {
-                char num[UTIL_PATH_SIZE];
-
-                sprintf(num, "%u", minor(udev_device_get_devnum(dev)));
-                l = strpcpy(&s, l, num);
+        case SUBST_MAJOR:
+        case SUBST_MINOR:
+                r = sd_device_get_devnum(dev, &devnum);
+                if (r < 0 && r != -ENOENT)
+                        return r;
+                if (asprintf(&buf, "%u", r < 0 ? 0 : entry->type == SUBST_MAJOR ? major(devnum) : minor(devnum)) < 0)
```

Comment:
Hmm, the switch from a stack buffer to heap allocation here is not nice. Potentially this function can be called quite often, and people already complain that udev rule processing is slow. What about using DECIMAL_STRING_WIDTH and `xsprintf` here?